### PR TITLE
tests: poll HeartbeatResponse

### PR DESCRIPTION
**Pull request by overvenus**
In tests, the mock pd-server(TestPdClient::Cluster) does not push region heartbeat responses. This PR try to simulate pushing by polling responses in the background. 

It's part of region heartbeat cache #2701, for fixing tests.

### CODE REVIEW

**Comment by overvenus** [comment]
PTAL

File: tests/raftstore/pd.rs [hunk 1]
```diff
@@ -56,18 +51,121 @@ impl Default for Store {
...
+            },
+            SchedulePolicy::TillSuccess => true,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+enum Operator {
+    AddPeer {
+        peer: Either<metapb::Peer, metapb::Peer>,
```
Comment (BusyJay) [inline tests/raftstore/pd.rs]: Add a comment explaining what are they respectively.

File: tests/raftstore/pd.rs [hunk 2]
```diff
@@ -56,18 +51,121 @@ impl Default for Store {
...
+                }
+            }
+            Operator::RemovePeer { ref peer, .. } => {
+                new_pd_change_peer(eraftpb::ConfChangeType::RemoveNode, peer.clone())
+            }
+            Operator::TransferLeader { ref peer, .. } => new_pd_transfer_leader(peer.clone()),
+        }
+    }
+
+    fn finished(
```
Comment (BusyJay) [inline tests/raftstore/pd.rs]: s/finished/try_finished/

File: tests/raftstore/pd.rs [hunk 3]
```diff
@@ -56,18 +51,121 @@ impl Default for Store {
...
     region_sizes: HashMap<u64, u64>,
     base_id: AtomicUsize,
-    rule: Option<Rule>,
 
     store_stats: HashMap<u64, pdpb::StoreStats>,
     split_count: usize,
 
+    // region id -> Operator
+    operators: HashMap<u64, Operator>,
+    enable_max_peer_count_check: bool,
```
Comment (BusyJay) [inline tests/raftstore/pd.rs]: s/_max//

File: tests/raftstore/pd.rs [hunk 4]
```diff
@@ -505,39 +664,50 @@ impl TestPdClient {
...
-                region_id, peer, region
-            );
-            if region.get_id() != region_id {
-                return None;
-            }
-            new_pd_add_change_peer(region, peer.clone())
-        });
+        let op = Operator::AddPeer {
+            peer: Either::Left(peer),
+            policy: SchedulePolicy::Repeat(1),
```
Comment (BusyJay) [inline tests/raftstore/pd.rs]: Why use 1?

**Comment by overvenus** [comment]
A friendly ping.

File: tests/raftstore/pd.rs (hunk 4 above)
Comment (BusyJay) [reply tests/raftstore/pd.rs]: In previous implementation, `AddPeer` is not only scheduled once.

File: tests/raftstore/pd.rs [hunk 5]
```diff
@@ -325,13 +422,61 @@ impl Cluster {
...
+            return Some(Operator::RemovePeer { peer, policy });
+        }
 
-            change_peer.set_change_type(eraftpb::ConfChangeType::RemoveNode);
-            change_peer.set_peer(region.get_peers()[pos].clone());
-            resp.set_change_peer(change_peer);
+        None
+    }
+
+    fn poll_heartbeat_responses_of(&mut self, store_id: u64) -> Vec<pdpb::RegionHeartbeatResponse> {
```
Comment (BusyJay) [inline tests/raftstore/pd.rs]: s/_of/_for/

File: tests/raftstore/pd.rs (hunk 4 above)
Comment (overvenus) [reply tests/raftstore/pd.rs]: Yes. In previous,  it's scheduled when TiKV heartbeats PD periodically. Now we schedule it on every polling. So they are essentially the same.

Comment (BusyJay) [reply tests/raftstore/pd.rs]: But the policy here is `Repeat(1)`, how can they be the same?

Comment (huachaohuang) [reply tests/raftstore/pd.rs]: So, it seems you should replace `add_peer` and `remove_peer` with `SchedulePolicy::TillSuccess`, and leave `must_add_peer` and `must_remove_peer` unchanged.

File: tests/raftstore/pd.rs [hunk 6]
```diff
@@ -505,39 +669,50 @@ impl TestPdClient {
         let region = self.get_region_by_id(region_id).wait().unwrap();
         panic!("region {:?} has peer {:?}", region, peer);
     }
+
     pub fn add_region(&self, region: &metapb::Region) {
         self.cluster.wl().add_region(region)
     }
 
+    pub fn ensure_transfer_leader(&self, region_id: u64, peer: metapb::Peer) {
```
Comment (huachaohuang) [inline tests/raftstore/pd.rs]: Don't need to add `ensure_` if all these operators use `SchedulePolicy::TillSuccess`, like `add_peer` and `remove_peer`.

File: tests/raftstore_cases/test_snap.rs [hunk 7]
```diff
@@ -449,7 +449,7 @@ fn test_snapshot_with_append<T: Simulator>(cluster: &mut Cluster<T>) {
         .sim
         .wl()
         .add_recv_filter(4, box SnapshotAppendFilter::new(tx));
-    pd_client.add_peer(1, new_peer(4, 5));
+    pd_client.must_add_peer(1, new_peer(4, 5));
```
Comment (huachaohuang) [inline tests/raftstore_cases/test_snap.rs]: Seems not necessary to use `must_add_peer`.

**Comment by BusyJay** [comment]
Any updates?

File: tests/raftstore/pd.rs (hunk 4 above)
Comment (overvenus) [reply tests/raftstore/pd.rs]: Some tests expect add-peer failure due to unsafe conf change, so I think repeat 1 is enough, `TillSuccess` will repeat the operator till it is success.

**Comment by overvenus** [comment]
A friendly ping.

File: tests/raftstore/pd.rs [hunk 8]
```diff
@@ -505,39 +669,50 @@ impl TestPdClient {
...
+        let op = Operator::RemovePeer {
+            peer,
+            policy: SchedulePolicy::TillSuccess,
+        };
+        self.schedule_operator(region_id, op);
     }
 
     pub fn must_remove_peer(&self, region_id: u64, peer: metapb::Peer) {
-        self.remove_peer(region_id, peer.clone());
+        let op = Operator::RemovePeer {
```
Comment (huachaohuang) [inline tests/raftstore/pd.rs]: This is the same as `remove_peer`.

File: tests/raftstore/pd.rs [hunk 9]
```diff
@@ -505,39 +669,50 @@ impl TestPdClient {
...
+        let op = Operator::AddPeer {
+            peer: Either::Left(peer),
+            policy: SchedulePolicy::TillSuccess,
+        };
+        self.schedule_operator(region_id, op);
     }
 
     pub fn must_add_peer(&self, region_id: u64, peer: metapb::Peer) {
-        self.add_peer(region_id, peer.clone());
+        let op = Operator::AddPeer {
```
Comment (huachaohuang) [inline tests/raftstore/pd.rs]: This is the same as `add_peer`.

Comment (overvenus) [reply tests/raftstore/pd.rs]: But it waits the result.

File: tests/raftstore/pd.rs [hunk 10]
```diff
@@ -56,18 +51,123 @@ impl Default for Store {
     }
 }
 
+#[derive(Debug, Clone, PartialEq)]
+enum SchedulePolicy {
+    /// Repeat a Operator.
```
Comment (huachaohuang) [inline tests/raftstore/pd.rs]: s/a/an

**Comment by hicqu** [comment]
LGTM.

**Pull request merged**